### PR TITLE
QUIC: Add quic event driver for QUICFrameGenerator 

### PR DESCRIPTION
**Pull request by scw00**

### CODE REVIEW

File: iocore/net/P_QUICNetVConnection.h [hunk 1]
```diff
@@ -145,6 +147,11 @@ class QUICNetVConnection : public UnixNetVConnection,
   void init(QUICConnectionId peer_cid, QUICConnectionId original_cid, QUICConnectionId first_cid, UDPConnection *,
             QUICPacketHandler *, QUICConnectionTable *ctable);
 
+  // QUICEventDriver
+  void set_write_event_pending(bool pend = true) override;
```
Comment (scw00) [inline iocore/net/P_QUICNetVConnection.h]: Pending sending processing when credit is not enough. The `generator` should set pending flag when it is pended by credit. So the credit manager(CC, FC) can reenable write when credit frame coming (MAX_XXX_FRAME or ACK FRAME).

File: iocore/net/P_QUICNetVConnection.h [hunk 2]
```diff
@@ -145,6 +147,11 @@ class QUICNetVConnection : public UnixNetVConnection,
...
+  bool is_event_pending() const;
```
Comment (scw00) [inline iocore/net/P_QUICNetVConnection.h]: Pending sending processing when credit is not enough. The generator should set pending flag when it is pended by credit. So the credit manager(CC, FC) can reenable write when credit frame coming (MAX_XXX_FRAME or ACK FRAME).

File: iocore/net/P_QUICNetVConnection.h [hunk 3]
```diff
@@ -145,6 +147,11 @@ class QUICNetVConnection : public UnixNetVConnection,
...
+  void reenable_write();
```
Comment (scw00) [inline iocore/net/P_QUICNetVConnection.h]: `generator` should reenable_write when it wants to send frames.

**Comment by maskit** [comment]
For this one, you may want to consult with @masaori335 and @vmamidi because they're working on WRITE_READY event for H2.

**Comment by scw00** [comment]
[skip ci]

**Comment by maskit** [comment]
> Pending sending processing when credit is not enough. The generator should set pending flag when it is pended by credit. So the credit manager(CC, FC) can reenable write when credit frame coming (MAX_XXX_FRAME or ACK FRAME).

I think I get what you are trying to do, but not sure why we should do so. Can you tell us a bit more detail? What is the issue?

**Comment by scw00** [comment]
Because we schedule write_packet_ready all the time. And it spend lost of cpu cycles for generators which are not ready to generate frame. 

For performance, the MAX_XX_DATA_FRAME will be blocked sometimes. So it greatly decreases the sending latency.

**Comment by maskit** [comment]
> Because we schedule write_packet_ready all the time. And it spend lost of cpu cycles for generators which are not ready to generate frame.

That's very true, but what we need to fix is infinite write_packet_ready. We haven't been able to fix it, but a consensus made at the last QUIC Hackathon was:
> ### Why WRITE_READY is scheduled by NetVC itself?
> Because there is only one FD
> -> Delayed bind should resolve the WRITE_READY spin issue

image

**Comment by scw00** [comment]
> Delayed bind should resolve the WRITE_READY spin issue

what does this means ? can you please share more information ?

**Comment by maskit** [comment]
I believe "delayed bind" creates a FD and we will be able to epoll somehow. Then it should schedule WRITE_READY when socket is really writable. It might not be true, but an important point is that we think PACKET_WRITE_READY spin is a bad idea and it should be eliminated.

**Comment by maskit** [comment]
So I think introducing something based on packet_write_ready spin design is not a great idea (we would need to remove two things in the future, the spin and the new mechanism based on it).

**Comment by scw00** [comment]
Thanks @maskit . And let's focus on delay bind implementation.

**Comment by maskit** [comment]
I've just remembered how we are going to remove WRITE_READY spin. #5352

**Pull request closed without merging**